include/alpaka/core/UniqueId.hpp

```diff
@@ -0,0 +1,152 @@
+* it under the terms of the GNU Lesser General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* alpaka is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with alpaka.
+* If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <boost/predef.h>           // workarounds
+
+#include <cstddef>                  // std::size_t
+
+namespace alpaka
+{
+    namespace core
+    {
+        namespace detail
+        {
+            //#############################################################################
+            // Based on the code from Filip Ros�en at http://b.atch.se/posts/constexpr-counter/
+            //#############################################################################
+            /*template<
+                std::size_t N>
+            struct flag;
+
+            //-----------------------------------------------------------------------------
+            //
+            //-----------------------------------------------------------------------------
+            template<
+                std::size_t N>
+            constexpr int adl_flag(flag<N>);*/
+
+            //#############################################################################
+            //
+            //#############################################################################
+            template<
+                std::size_t N>
+            struct flag
+            {
+                friend constexpr std::size_t adl_flag(flag<N>);
+            };
+
+            //#############################################################################
+            //
+            //#############################################################################
+            template<
+                std::size_t N>
+            struct writer
+            {
+                //-----------------------------------------------------------------------------
+                //
+                //-----------------------------------------------------------------------------
+                friend constexpr std::size_t adl_flag(flag<N>)
+                {
+                    return N;
+                }
+
+                static constexpr std::size_t value = N;
+            };
+
+#ifdef BOOST_COMP_MSVC
+            //-----------------------------------------------------------------------------
+            //! The matcher.
+            //-----------------------------------------------------------------------------
+            template<
+                std::size_t N,
+                class = char[noexcept(adl_flag(flag<N>{})) ? +1u : -1u]>
```

Comment:
The current problem with NVCC is that it not supports `noexcept` on this way we need it here. NVCC implements [this](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2010/n3050.html#operator-noexcept-expression).
The second pass in line 92 is also not working with nvcc :-(